### PR TITLE
ENH: allow approx_fprime to work with vector-valued func

### LINKED ISSUE #2993: optimize.approx_fprime & jacobians
Hello,
While trying to run scipy.optimize.check_grad  function in the following code
# !/usr/bin/python

'''Fitting the function y = c + r*sin(phi + alpha)'''
from math import sin, cos, pi
import scipy.optimize as op
import numpy as np

def f(als, c=1., r=1., phi=15*pi/180.):
    return np.array([y(a, c, r, phi) for a in als])

def y(a,  c=1., r=1., phi=15*pi/180.):
    return c + r \* sin(phi + a)

def y_grad(als, c=1., r=1., phi=15*pi/180):
    return np.array([y_prime(a, c, r, phi) for a in als])

def y_prime(a, c=1., r=1., phi=15*pi/180):
    return -r \* cos(phi + a)

def diffsquares(y_exp, y_fit):
    return (y_exp - y_fit)**2

step = 2_pi / 360.
a = np.arange(0., 2_pi + step, step)

result = op.check_grad(f, y_grad, a)

print 'result', result

on Ubuntu 12.04 (scipy version '0.9.0') I get the following error:

Traceback (most recent call last):
  File "./fit.py", line 28, in <module>
    result = op.check_grad(f, y_grad, a)
  File "/usr/lib/python2.7/dist-packages/scipy/optimize/optimize.py", line 382, in check_grad
    return sqrt(sum((grad(x0,_args)-approx_fprime(x0,func,_epsilon,_args))*_2))
  File "/usr/lib/python2.7/dist-packages/scipy/optimize/optimize.py", line 377, in approx_fprime
    grad[k] = (f(_((xk+ei,)+args)) - f0)/epsilon
ValueError: setting an array element with a sequence.

looking in the source code I think there is an obvious typo there, the numerator is not indexed, I have checked the code on gitHub and the bug seems to be still present (line 610 https://github.com/scipy/scipy/blob/master/scipy/optimize/optimize.py) .

line 610
grad[k] = (f(*((xk + d,) + args)) - f0) / d[k]

should probably read
grad[k] = (f(*((xk + d,) + args))[k] - f0[k]) / d[k]

Cheers,
Martin

### LINKED ISSUE #6992: optimize.check_grad does not work for complex-valued functions
The function ``optimize.check_grad`` does not work for complex-valued functions.

Here is an adapted example from the [documentation of check_grad](http://scipy.github.io/devdocs/generated/scipy.optimize.check_grad.html):

    def func(x):
        return x[0]**2 - 0.5j * x[1]**3
    
    def grad(x):
        return [2 * x[0], -1.5j * x[1]**2]
    
    from scipy.optimize import check_grad
    print(check_grad(func, grad, [1.5, -1.5]))

Expected: a (real) float close to zero.
Got: a ComplexWarning is raised and the returned value is non-zero.

This should be solved by replacing the last line:

    return sqrt(sum((grad(x0, *args) - approx_fprime(x0, func, step, *args))**2))

by:

    return sqrt(sum(numpy.abs((grad(x0, *args) - approx_fprime(x0, func, step, *args))**2))) 

(take the absolute value)

Edit : A long term solution would be using optimize._numdiff.check_derivative instead. However the function optimize._numdiff._dense_difference has to be adapted to complex valued function too.

### PR DESCRIPTION
**Pull request by andyfaff**
Closes #2993 and closes #6992.

This would allow `approx_fprime` to work with vector-valued functions.

- All the PR does is remove the check on `func` returning a scalar. The differentiation is already done by `_numdiff.approx_derivative`.
- `check_grad` is amended slightly to work for vector valued functions, but not along random projections.
- The docstrings are amended.

What's the best way of adding a changelog to the docstring to let people know of the updated behaviour?

### CODE REVIEW

**Comment by ev-br** [comment]
> What's the best way of adding a changelog to the docstring to let people know of the updated behaviour?

Not sure about the best way; one way is to use the ``.. versionchanged::`` directive maybe.

**Comment by tupui** [comment]
Actually there is a directive for that: `.. versionchanged:: version`

**Comment by andyfaff** [comment]
I used the versionchanged RST directive, I think it looks ok in the doc artefact. I fixed the failing test (needed a conversion to an array). I think the current test fails are unrelated.

File: scipy/optimize/_optimize.py [hunk 1]
```diff
@@ -908,17 +908,29 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
...
+        Function of which to estimate the derivatives of. Has the signature
+        ``f(xk, *args)`` where `xk` is the argument in the form of a 1-D array
+        and `args` is a  tuple of any additional fixed parameters needed to
+        completely specify the function. The argument xk passed to this
+        function is ndarray of shape (n,) (never a scalar even if n=1).
+        It must return 1-D array_like of shape (m,) or a scalar.
+
+        .. versionchanged:: 1.8.0
+            `f` is now able to return a 1-D array-like, with the (m, n)
+            Jacobian being estimated.
```
Comment (tupui) [inline scipy/optimize/_optimize.py:932]:
```suggestion
        .. versionchanged:: 1.8.0
           `f` is now able to return a 1-D array-like, with the
           :math:`(m, n)` Jacobian being estimated.
```

File: scipy/optimize/_optimize.py [hunk 2]
```diff
@@ -1064,8 +1070,9 @@ def g(w, func, x0, v, *args):
         raise ValueError("{} is not a valid string for "
                          "``direction`` argument".format(direction))
 
-    return sqrt(sum((analytical_grad -
-                     approx_fprime(vars, _func, step, *_args))**2))
+    return np.sqrt(np.sum(np.abs(
+        (analytical_grad - approx_fprime(vars, _func, step, *_args))**2
+    )))
```
Comment (tupui) [inline scipy/optimize/_optimize.py:1075]: If we have a vector, don't we want to have the possibility to access the whole error vector?

**Review by tupui — COMMENTED** [review]
Thanks @andyfaff. I left a few suggestions.

File: scipy/optimize/_optimize.py (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/_optimize.py:1075]: From the existing docs:

> err : float
>     The square root of the sum of squares (i.e., the 2-norm) of the
>     difference between ``grad(x0, *args)`` and the finite difference
>     approximation of `grad` using func at the points `x0`.

The `check_grad` function is supposed to return a float by reducing the `(n,)` derivative array to a single value. This behaviour is unchanged, except it also does it for an (m, n) array.

Comment (tupui) [reply scipy/optimize/_optimize.py:1076]: I was not clear, my question was more: should we allow the user to request a vector? Like some metric in scikit-learn were you can request a scalar or vector (and for some you can also specify how the scalar is computed along the vector)

**Comment by andyfaff** [comment]
@tupui, is this mergeable as is, it would close a couple of issues.

File: scipy/optimize/_optimize.py [hunk 3]
```diff
@@ -908,17 +908,30 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
...
-        supplied in ``*args``. Should return a scalar, the value of the
-        function at `xk`.
+        Function of which to estimate the derivatives of. Has the signature
+        ``f(xk, *args)`` where `xk` is the argument in the form of a 1-D array
+        and `args` is a tuple of any additional fixed parameters needed to
+        completely specify the function. The argument `xk` passed to this
+        function is an ndarray of shape (n,) (never a scalar even if n=1).
+        It must return a 1-D array_like of shape (m,) or a scalar.
+
+        .. versionchanged:: 1.8.0
```
Comment (tupui) [inline scipy/optimize/_optimize.py:931]:
```suggestion
        .. versionchanged:: 1.9.0
```

**Review by tupui — APPROVED** [review]
> @tupui, is this mergeable as is, it would close a couple of issues.

It is, yes. Let's merge it now if it unlocks things (I am just adjusting the version flag to 1.9), the suggestion/question I had can be dealt with later. I will let a few CI go green in case anything else changed in the meantime and merge.

**Pull request merged**